KAFKA-15853: Move configDef out of core

OmniaGM opened the pull request:
Moving configDef from core into `server::KafkaConfig`, keeping the old reference from `core::KafkaConfig` for the time being to avoid the confusion when we have need to import `server::KafkaConfig` and `core::KafkaConfig`.  

The following PRs will move the rest of core::KafkaConfig`

This will reduce conflicts while finishing the last part of moving KafkaConfig out of core
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

OmniaGM committed:
KAFKA-15853: Move configDef out of KafkaConfig

chia7712 commented:
@OmniaGM Could you please fix the conflicts

OmniaGM committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-15853-move-configDef-out

OmniaGM committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-15853-move-configDef-out

chia7712 commented:
@OmniaGM Please fix the conflicts, thanks!

chia7712 commented:
@OmniaGM Sorry that please fix the conflicts again :_

OmniaGM committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-15853-move-configDef-out

nizhikov commented:
There are more conflicts now.

@OmniaGM Can you, please, fix it.

OmniaGM committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-15853-move-configDef-out

OmniaGM commented:
Fixed let's hope this is the last one :)

chia7712 commented:
> Fixed let's hope this is the last one :)

if it can go conflicted, it will :)

chia7712 commented on server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java:
```diff
@@ -0,0 +1,420 @@
+import static org.apache.kafka.common.config.ConfigDef.Type.SHORT;
+import static org.apache.kafka.common.config.ConfigDef.Type.STRING;
+
+public class KafkaConfig {
```
Maybe we can declare a enum type to distinguish those huge configs? For example:
```java
public enum KafkaConfigDef {
  REMOTE_LOG_MANAGER(RemoteLogManagerConfig.CONFIG_DEF),
  ZOOKEEPER(new ConfigDef().define(ZkConfigs.ZK_CONNECT_CONFIG, STRING, null, HIGH, ZkConfigs.ZK_CONNECT_DOC)
      .define(ZkConfigs.ZK_SESSION_TIMEOUT_MS_CONFIG, INT, ZkConfigs.ZK_SESSION_TIMEOUT_MS, HIGH, ZkConfigs.ZK_SESSION_TIMEOUT_MS_DOC)
      .define(ZkConfigs.ZK_CONNECTION_TIMEOUT_MS_CONFIG, INT, null, HIGH, ZkConfigs.ZK_CONNECTION_TIMEOUT_MS_DOC)
      .define(ZkConfigs.ZK_ENABLE_SECURE_ACLS_CONFIG, BOOLEAN, ZkConfigs.ZK_ENABLE_SECURE_ACLS, HIGH, ZkConfigs.ZK_ENABLE_SECURE_ACLS_DOC)
      .define(ZkConfigs.ZK_MAX_IN_FLIGHT_REQUESTS_CONFIG, INT, ZkConfigs.ZK_MAX_IN_FLIGHT_REQUESTS, atLeast(1), HIGH, ZkConfigs.ZK_MAX_IN_FLIGHT_REQUESTS_DOC)
      .define(ZkConfigs.ZK_SSL_CLIENT_ENABLE_CONFIG, BOOLEAN, ZkConfigs.ZK_SSL_CLIENT_ENABLE, MEDIUM, ZkConfigs.ZK_SSL_CLIENT_ENABLE_DOC)
      .define(ZkConfigs.ZK_CLIENT_CNXN_SOCKET_CONFIG, STRING, null, MEDIUM, ZkConfigs.ZK_CLIENT_CNXN_SOCKET_DOC)
      .define(ZkConfigs.ZK_SSL_KEY_STORE_LOCATION_CONFIG, STRING, null, MEDIUM, ZkConfigs.ZK_SSL_KEY_STORE_LOCATION_DOC)
      .define(ZkConfigs.ZK_SSL_KEY_STORE_PASSWORD_CONFIG, PASSWORD, null, MEDIUM, ZkConfigs.ZK_SSL_KEY_STORE_PASSWORD_DOC)
      .define(ZkConfigs.ZK_SSL_KEY_STORE_TYPE_CONFIG, STRING, null, MEDIUM, ZkConfigs.ZK_SSL_KEY_STORE_TYPE_DOC)
      .define(ZkConfigs.ZK_SSL_TRUST_STORE_LOCATION_CONFIG, STRING, null, MEDIUM, ZkConfigs.ZK_SSL_TRUST_STORE_LOCATION_DOC)
      .define(ZkConfigs.ZK_SSL_TRUST_STORE_PASSWORD_CONFIG, PASSWORD, null, MEDIUM, ZkConfigs.ZK_SSL_TRUST_STORE_PASSWORD_DOC)
      .define(ZkConfigs.ZK_SSL_TRUST_STORE_TYPE_CONFIG, STRING, null, MEDIUM, ZkConfigs.ZK_SSL_TRUST_STORE_TYPE_DOC)
      .define(ZkConfigs.ZK_SSL_PROTOCOL_CONFIG, STRING, ZkConfigs.ZK_SSL_PROTOCOL, LOW, ZkConfigs.ZK_SSL_PROTOCOL_DOC)
      .define(ZkConfigs.ZK_SSL_ENABLED_PROTOCOLS_CONFIG, LIST, null, LOW, ZkConfigs.ZK_SSL_ENABLED_PROTOCOLS_DOC)
      .define(ZkConfigs.ZK_SSL_CIPHER_SUITES_CONFIG, LIST, null, LOW, ZkConfigs.ZK_SSL_CIPHER_SUITES_DOC)
      .define(ZkConfigs.ZK_SSL_ENDPOINT_IDENTIFICATION_ALGORITHM_CONFIG, STRING, ZkConfigs.ZK_SSL_ENDPOINT_IDENTIFICATION_ALGORITHM, LOW, ZkConfigs.ZK_SSL_ENDPOINT_IDENTIFICATION_ALGORITHM_DOC)
      .define(ZkConfigs.ZK_SSL_CRL_ENABLE_CONFIG, BOOLEAN, ZkConfigs.ZK_SSL_CRL_ENABLE, LOW, ZkConfigs.ZK_SSL_CRL_ENABLE_DOC)
      .define(ZkConfigs.ZK_SSL_OCSP_ENABLE_CONFIG, BOOLEAN, ZkConfigs.ZK_SSL_OCSP_ENABLE, LOW, ZkConfigs.ZK_SSL_OCSP_ENABLE_DOC));

  private final ConfigDef configDef;

  KafkaConfigDef(ConfigDef configDef) {
    this.configDef = configDef;
  }

  public static ConfigDef all() {
    ConfigDef all = new ConfigDef();
    Arrays.stream(KafkaConfigDef.values()).forEach(e -> e.configDef.configKeys().values().forEach(all::define));
    return all;
  }
}
```

That is more readable to me. WDYT?

OmniaGM replied:
Sorry for the late response. It would be nice separation however I am not sure it would be useful as these definition rarely will be re-used with the same structure and definition outside of their classes. For example `RemoteLogManagerConfig.CONFIG_DEF` will be only used in `RemoteLogManagerConfig` and not needed outside of this context. The `ZOOKEEPER` is only used in KafkaConfig and will not be used outside of this. 

Also we have some configs that common between server and other components of kafka and having multiple definition for different context for the same configs might be confusing going forward. 

We also might create confusion regarding where to define these config definitions inside  `Config/Configs` class or within `KafkaConfigDef`. For example we can move most config definition for things like `QUOTA`, `GROUP_COORDINATOR`,  `REPLICATION` etc to their `Configs` classes however things like `SSL`, `SASL` get mostly populated from public classes and `LOG` while most of the `CONFIG` fields are defined in private class `ServerLogConfigs`, the defaults are defined in storage package and we can't depend on the storage as it will create circular dependency between `storage` and `server-common`

chia7712 replied:
you are right. my previous idea is over-engineering :_

chia7712 commented on server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java:
```diff
@@ -0,0 +1,420 @@
+import static org.apache.kafka.common.config.ConfigDef.Type.SHORT;
+import static org.apache.kafka.common.config.ConfigDef.Type.STRING;
+
+public class KafkaConfig {
```
If it is used to return `ConfigDef` only, maybe it can be renamed to `KafkaConfigDefs`?

OmniaGM replied:
I was planning to move the rest of KafkaConfig to it and make this class an abstract in the next PR and KafkaConfig.scala will extend it for now until I move everything out

chia7712 replied:
> I was planning to move the rest of KafkaConfig to it and make this class an abstract in the next PR and KafkaConfig.scala will extend it for now until I move everything out

not sure whether I catch your point. it seems we will have a java version KafkaConfig but it is smaller?

OmniaGM replied:
I was planning to keep 2 KafkaConfig for the time being while moving everything out of core. 
Using `KafkaConfigDef` has the following 
**Pros:** 
-  The ConfigDef is huge so having this out of `KafkaConfig` whether it is the original `KafkaConfig` in `core` or the new `KafkaConfig` in `server`. 
**Cons:** 
- We will have class that only contains 1 static variable and most likely not going to scale for any thing else. 

Having 2 `KafkaConfig` (most likely the one in core will keep getting smaller and smaller till we are ready to delete). The idea I have is  the one in `server` will be an abstract class that will be extended by the one in `core` and I slowly will move stuff out of `core::KafkaConfig` to `server::KafkaConfig`
**Pros:** 
- easier to review as all references and imports will point to KafkaConfig in core during the process of moving the getters into `server`. And the last PR will be just changing the import to the right `KafkaConfig`
- Removing ZK code from `core::KafkaConfig` (which will happened in the next few weeks) will not create conflicts in `server::KafkaConfig` as I can ignore moving ZK `getters` and validation. 
**Cons:**
- We have two `KafkaConfig` for a period of time


I don't mind renaming this to `KafkaConfigDef` but my point is going forward  we most likely going to have 2 `KafkaConfig` for a short period of time.

chia7712 replied:
It seems option2 is better. My concern was that `KafkaConfig` will be still mixed with all configs even though we move it to server module.

chia7712 replied:
BTW, please fix the conflicts again :)

OmniaGM replied:
> It seems option2 is better. My concern was that KafkaConfig will be still mixed with all configs even though we move it to server module.

Unfornatly yes. I was planning to make `server::KafkaConfig` as abstract for the time being until I move everything out.

OmniaGM replied:
> BTW, please fix the conflicts again :)

Fixed now

chia7712 replied:
>  I was planning to make server::KafkaConfig as abstract for the time being until I move everything out.

That is a intermediate state, right?  I assume there will be specific config calss for different component (such as `RemoteLogManagerConfig` and `QuorumConfig`). server::KafkaConfig will be either nonexistent or small class (maybe do check for all configs or run once in starting)

OmniaGM replied:
This would be another piece of work. Right now in the context of [KAFKA-15853](https://github.com/apache/kafka/pull/16116#top) I want to move `KafkaConfig` out of core to unblock the rest of [KAFKA-15852](https://issues.apache.org/jira/browse/KAFKA-15852) that depends on `KafkaConfig`

chia7712 replied:
Got it.

chia7712 commented on server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java:
```diff
@@ -0,0 +1,420 @@
+
+public class KafkaConfig {
+    @SuppressWarnings("deprecation")
+    public final static ConfigDef CONFIG_DEF =  new ConfigDef()
```
Could it include `RemoteLogManagerConfig.configDef()`?

chia7712 reviewed (commented):
@OmniaGM thanks for updated PR!

OmniaGM committed:
add RemoteLogManagerConfig.CONFIG_DEF

OmniaGM committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-15853-move-configDef-out

chia7712 commented on server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java:
```diff
@@ -0,0 +1,421 @@
+import static org.apache.kafka.common.config.ConfigDef.Type.SHORT;
+import static org.apache.kafka.common.config.ConfigDef.Type.STRING;
+
+public class KafkaConfig {
```
Could you please add comments to explain why we have "two" KafkaConfig in code base? That can avoid confusion when someone try to add something to `KafkaConfig` in the future.

OmniaGM committed:
add commit to explain the diff between the two kafka configs
- Make org.apache.kafka.server.config.KafkaConfig abstract so others can start move their methods there

chia7712 commented on server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java:
```diff
@@ -0,0 +1,432 @@
+ * Any code depends on kafka.server.KafkaConfig will keep for using kafka.server.KafkaConfig for the time being until we move it out of core
+ * For more details check KAFKA-15853
+ */
+abstract public class KafkaConfig extends AbstractConfig {
```
It seems `public abstract class` is used more frequent than `abstract public class`. Maybe we can align it?

chia7712 reviewed (commented):
@OmniaGM thanks for updated PR. could you please fix build error?
```
[2024-06-10T15:13:31.280Z] > Task :core:compileScala

[2024-06-10T15:13:31.280Z] [Error] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-16116/core/src/main/scala/kafka/server/KafkaConfig.scala:27:40: Unused import
```

OmniaGM committed:
feedback

chia7712 commented:
@OmniaGM Sorry that conflicts happen again :_

OmniaGM committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-15853-move-configDef-out

OmniaGM commented:
> @OmniaGM Sorry that conflicts happen again :_

Done let's hope this is the final one :D

OmniaGM committed:
fix

mimaison commented on server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java:
```diff
@@ -0,0 +1,447 @@
+        .define(KRaftConfigs.MIGRATION_ENABLED_CONFIG, BOOLEAN, false, HIGH, KRaftConfigs.MIGRATION_ENABLED_DOC)
+        .define(KRaftConfigs.ELR_ENABLED_CONFIG, BOOLEAN, false, HIGH, KRaftConfigs.ELR_ENABLED_DOC)
+        .defineInternal(KRaftConfigs.MIGRATION_METADATA_MIN_BATCH_SIZE_CONFIG, INT, KRaftConfigs.MIGRATION_METADATA_MIN_BATCH_SIZE_DEFAULT, atLeast(1),
+                MEDIUM, KRaftConfigs.MIGRATION_METADATA_MIN_BATCH_SIZE_DOC)
```
Why do we have a newline here?

OmniaGM replied:
No other reason than it is longer line than the rest

OmniaGM replied:
removed the new lines

mimaison reviewed (approved):
LGTM
I left a couple of suggestions but these can be addressed in one of the follow up PRs you have for KafkaConfig

mimaison commented on server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java:
```diff
@@ -0,0 +1,447 @@
+public abstract class KafkaConfig extends AbstractConfig {
+    @SuppressWarnings("deprecation")
+    public static final ConfigDef CONFIG_DEF =  new ConfigDef(RemoteLogManagerConfig.configDef())
+        /** ********* Zookeeper Configuration ***********/
```
We can replace these javadoc-style comments with regular comments

OmniaGM replied:
done

mimaison reviewed (approved):
LGTM
I left a couple of suggestions but these can be addressed in one of the follow up PRs you have for KafkaConfig

OmniaGM committed:
address feedback

OmniaGM committed:
rename org.apache.kafka.server.config.KafkaConfig to AbstractKafkaConfig

OmniaGM commented:
I have renamed the class to `AbstractKafkaConfig` to get ripped of `NM_SAME_SIMPLE_NAME_AS_SUPERCLASS` error from spotbug

OmniaGM committed:
Break CONFIG_DEF into multi def each one defined in its own Config(s) class

OmniaGM commented:
>That is a intermediate state, right? I assume there will be specific config calss for different component (such as RemoteLogManagerConfig and QuorumConfig). server::KafkaConfig will be either nonexistent or small class (maybe do check for all configs or run once in starting)

Looking into this again. I think we can squeeze this in as it will reduce conflict with this work going forward specially with the ongoing work for shared groups and new coordinator.

mimaison commented:
@OmniaGM `spotlessJavaCheck` is failing for `transaction-coordinator` and `server-common`: https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-16116/17/pipeline

OmniaGM committed:
fix pipeline

chia7712 commented:
> I think we can squeeze part of this vision in as it will reduce conflict with this work going forward specially with the ongoing work for shared groups and new coordinator which I want to avoid as it is hardest to fix

I love this code style

> please have a look again. I would really appreciate if we can merge this pr as it is getting painful to keep up with the conflicts :)

I've set the alarm for this PR :)

chia7712 reviewed (approved):
LGTM

chia7712 commented:
wait a minute. it seems there are failed tests related to this PR

https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-16116/19/tests

chia7712 commented on transaction-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionLogConfigs.java:
```diff
@@ -51,4 +60,16 @@ public final class TransactionLogConfigs {
+
+            .define(TransactionLogConfigs.PRODUCER_ID_EXPIRATION_MS_CONFIG, INT, TransactionLogConfigs.PRODUCER_ID_EXPIRATION_MS_DEFAULT, atLeast(1), LOW, TransactionLogConfigs.PRODUCER_ID_EXPIRATION_MS_DOC)
+            // Configuration for testing only as default value should be sufficient for typical usage
+            .defineInternal(TransactionLogConfigs.PRODUCER_ID_EXPIRATION_CHECK_INTERVAL_MS_CONFIG, INT, TransactionLogConfigs.PRODUCER_ID_EXPIRATION_CHECK_INTERVAL_MS_DEFAULT, atLeast(1), LOW, TransactionLogConfigs.PRODUCER_ID_EXPIRATION_CHECK_INTERVAL_MS_DOC);
```
Please add `define(ServerLogConfigs.LOG_DIR_FAILURE_TIMEOUT_MS_CONFIG, LONG, ServerLogConfigs.LOG_DIR_FAILURE_TIMEOUT_MS_DEFAULT, atLeast(1), LOW, ServerLogConfigs.LOG_DIR_FAILURE_TIMEOUT_MS_DOC)` to fix the failed tests :)

OmniaGM replied:
done

OmniaGM replied:
I defined it in LogConfig.SERVER_CONFIG_DEF

OmniaGM committed:
add missing LOG_DIR_FAILURE_TIMEOUT_MS_CONFIG

mimaison reviewed (approved):
This is a bit of a weird state but I understand this is temporary and you have follow up PRs that should tidy things up.
LGTM, let's get this merged as it's a pain to keep in sync

mimaison commented:
None of the test failures are related, merging to trunk
